File: sample_code/backend/native-c/thales_protect_gateway_service.c
#include <pkcs11.h>
#include <stdio.h>
#include <string.h>

static const char *PROTECT_ROUTE = "/protectInput";
static const char *PROTECT_AND_CALL_LLM_ROUTE = "/protectInputAndCallLLM";
static const char *REVEAL_ROUTE = "/revealInput";

int ThalesProtectGatewayService(const char *route) {
    if (strcmp(route, PROTECT_ROUTE) == 0) {
        return 1;
    }
    if (strcmp(route, PROTECT_AND_CALL_LLM_ROUTE) == 0) {
        return 2;
    }
    if (strcmp(route, REVEAL_ROUTE) == 0) {
        return 3;
    }
    return 0;
}

int main(void) {
    printf("%d
", ThalesProtectGatewayService(PROTECT_ROUTE));
    return CKR_OK;
}

File: sample_code/backend/dotnet-api/Services/ThalesProtectGatewayService.cs
using Azure.Security.KeyVault.Cryptography;

namespace SampleCompany.Services;

public class ThalesProtectGatewayService
{
    private readonly CryptographyClient _cryptographyClient;
    private const string ProtectRoute = "/protectInput";
    private const string ProtectAndCallLlmRoute = "/protectInputAndCallLLM";
    private const string RevealRoute = "/revealInput";

    public ThalesProtectGatewayService(CryptographyClient cryptographyClient)
    {
        _cryptographyClient = cryptographyClient;
    }

    public string AuthenticateRequest(string authorizationHeader)
    {
        return string.IsNullOrWhiteSpace(authorizationHeader) ? "missing" : "ok";
    }

    public string ProtectInput() => ProtectRoute;

    public string ProtectInputAndCallLLM() => ProtectAndCallLlmRoute;

    public string RevealInput() => RevealRoute;
}
